XenModel/Actions/Updates/DownloadAndUpdateClientAction.cs

```diff
@@ -0,0 +1,286 @@
+        private Exception updateDownloadError;
+        private string checksum;
+        private WebClient client;
+
+        public string ByteProgressDescription { get; set; }
+
+        public DownloadAndUpdateClientAction(string updateName, Uri uri, string outputFileName, string checksum)
+            : base(null, string.Format(Messages.DOWNLOAD_CLIENT_INSTALLER_ACTION_TITLE, updateName),
+                string.Empty, true)
+        {
+            this.updateName = updateName;
+            address = uri;
+            downloadUpdate = address != null;
+            this.outputPathAndFileName = outputFileName;
+            this.checksum = checksum;
+        }
+
+        private void DownloadFile()
+        {
+            int errorCount = 0;
+            bool needToRetry = false;
+
+            client = new WebClient();
+            client.DownloadProgressChanged += client_DownloadProgressChanged;
+            client.DownloadFileCompleted += client_DownloadFileCompleted;
+
+            // register event handler to detect changes in network connectivity
+            NetworkChange.NetworkAvailabilityChanged += NetworkAvailabilityChanged;
+
+            try
+            {
+                do
+                {
+                    if (needToRetry)
+                        Thread.Sleep(SLEEP_TIME_BEFORE_RETRY_MS);
+
+                    needToRetry = false;
+
+                    client.Proxy = XenAdminConfigManager.Provider.GetProxyFromSettings(null, false);
+
+                    //start the download
+                    updateDownloadState = DownloadState.InProgress;
+                    client.DownloadFileAsync(address, outputPathAndFileName);
+
+                    bool updateDownloadCancelling = false;
+
+                    //wait for the file to be downloaded
+                    while (updateDownloadState == DownloadState.InProgress)
+                    {
+                        if (!updateDownloadCancelling && (Cancelling || Cancelled))
+                        {
+                            Description = Messages.DOWNLOAD_AND_EXTRACT_ACTION_DOWNLOAD_CANCELLED_DESC;
+                            client.CancelAsync();
+                            updateDownloadCancelling = true;
+                        }
+
+                        Thread.Sleep(SLEEP_TIME_TO_CHECK_DOWNLOAD_STATUS_MS);
+                    }
+
+                    if (updateDownloadState == DownloadState.Cancelled)
+                        throw new CancelledException();
+
+                    if (updateDownloadState == DownloadState.Error)
+                    {
+                        needToRetry = true;
+
+                        // this many errors so far - including this one
+                        errorCount++;
+
+                        // logging only, it will retry again.
+                        log.ErrorFormat(
+                            "Error while downloading from '{0}'. Number of errors so far (including this): {1}. Trying maximum {2} times.",

```

Comment:
👍 for making it clear that the log message error counts are inclusive of the current one.